test/distributed/fsdp/test_fsdp_use_orig_params.py

```diff
@@ -0,0 +1,875 @@
+        cuda_init_mode: CUDAInitMode,
+        init_optim_before_wrap: bool,
+        optim_class: Type[torch.optim.Optimizer],
+        multi_tensor: bool,
+        set_to_none: bool,
+        backward_prefetch: Optional[BackwardPrefetch],
+        cpu_offload: CPUOffload,
+        sharding_strategy: ShardingStrategy,
+    ):
+        """
+        Args:
+            init_optim_before_wrap (bool): If ``True``, initializes the
+                FSDP optimizer before wrapping the model with FSDP; otherwise,
+                initializes the FSDP optimizer after wrapping the model with
+                FSDP. We permit both forms of initialization to give users
+                flexibility.
+        """
+        if cuda_init_mode == CUDAInitMode.CUDA_AFTER and cpu_offload.offload_params:
+            return  # not supported
+        ddp_model, ddp_optim = self._get_ddp_transformer_and_optim(
+            optim_class=optim_class,
+            multi_tensor=multi_tensor,
+            find_unused_params=False,
+        )
+        fsdp_model, fsdp_optim = self._get_fsdp_transformer_and_optim(
+            cuda_init_mode=cuda_init_mode,
+            init_optim_before_wrap=init_optim_before_wrap,
+            optim_class=optim_class,
+            multi_tensor=multi_tensor,
+            sharding_strategy=sharding_strategy,
+            backward_prefetch=backward_prefetch,
+            cpu_offload=cpu_offload,
+        )
+        self._check_train_parity(
+            ddp_model, ddp_optim, fsdp_model, fsdp_optim, set_to_none
+        )
+
+    @skip_if_lt_x_gpu(2)
+    def test_diff_trainability(self):
+        """
+        Tests FSDP parity with DDP when using multiple parameter groups and
+        freezing the parameters in one parameter group.
+        """
+        self.run_subtests(
+            {
+                "multi_tensor": [False, True],
+                "sharding_strategy": [
+                    ShardingStrategy.FULL_SHARD,
+                    ShardingStrategy.SHARD_GRAD_OP,
+                    ShardingStrategy.NO_SHARD,
+                ],
+            },
+            self._test_diff_trainability,
+        )
+
+    def _test_diff_trainability(
+        self,
+        multi_tensor: bool,
+        sharding_strategy: ShardingStrategy,
+    ):
+        optim_class = torch.optim.Adam
+        ddp_model, ddp_optim = self._get_ddp_transformer_and_optim(
+            optim_class=optim_class,
+            multi_tensor=multi_tensor,
+            find_unused_params=True,
+        )
+        fsdp_model, fsdp_optim = self._get_fsdp_transformer_and_optim(
+            cuda_init_mode=CUDAInitMode.CUDA_BEFORE,
+            init_optim_before_wrap=False,
+            optim_class=optim_class,
+            multi_tensor=multi_tensor,
+            sharding_strategy=sharding_strategy,
+            backward_prefetch=BackwardPrefetch.BACKWARD_PRE,
+            cpu_offload=None,
+        )
+        # Freeze all biases (which happen to be in the same parameter group)
```

Comment:
what we're interested in is that if we have 2 params, both in same flat parameter, one can be frozen and other may not be, and this should work, right?